packages/loader/container-loader/src/container.ts

```diff
@@ -1327,16 +1325,14 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
             this.emit("connect", opsBehind);
 
             // Report telemetry after we set client id!
-            this.logConnectionStateChangeTelemetry(
-                ConnectionState.Connecting,
-                oldState,
-                "websocket established",
-                opsBehind);
-
-            if (deltaManager.connectionMode === "read") {
-                this.setConnectionState(
-                    ConnectionState.Connected,
-                    `joined as readonly`);
+            this.logConnectionStateChangeTelemetry(ConnectionState.Connecting, oldState);
+
+            // Check if we already processed our own join op through delta storage!
+            // we are fetching ops from storage in parallel to connecting to ordering service
+            // Given async processes, it's possible that we have already processed our own join message before
+            // connection was fully established.
+            if (this.getQuorum().has(details.clientId) || deltaManager.connectionMode === "read") {
+                this.setConnectionState(ConnectionState.Connected);
```

Comment:
Shouldn't we clear the pendingClientId after we are Connected? Does it serve any purpose to track it after we have a connection?
I see that we only clear it when we are Disconnected.

Reply:
We do that as part of setConnectionState() when disconnecting - this change does not change any of that logic.

Reply:
We can probably reset it on connected (based on my glancing through the code), but I'd rather keep these things separate (i.e. separate PR for easier back out in case of screwed up).

Reply:
Yes, I agree. We should not do this as part of this change. I will create an issue for this. Anyway, I don't think this is high pri.